### PR TITLE
Fix: #1837. Mention collisions.

### PR DESCRIPTION
**Pull request by ioggstream**
## This PR

Fix: #1837. Mention collisions.

### CODE REVIEW

**Review by jricher — CHANGES_REQUESTED** [review]
As mentioned in the linked issue, this PR is not complete: there needs to be a discussion of collisions and, more broadly, algorithm selection in the security considerations, since this draft leaves the choice of which algorithms to accept entirely up to the implementor. I would also expect to see a discussion of the danger of a receiver of Digest headers blindly accepting "all" algorithms defined in the draft or registry, even if they aren't deprecated. For example, a system in which all "good" clients use SHA512 but still accepts SHA256 is only really protected by SHA256 since that's the lightest weight item an attacker can make use of.

**Comment by ioggstream** [comment]
@martinthomson WDYT?

**Comment by martinthomson** [comment]
If you are talking about the standard assumptions about what a cryptographic hash function provides, you need to talk about what properties you are getting (or, conversely, what the not-secure options don't provide).  So not just collisions, but first- and second-preimage resistance.

Justin also points out that you need to consider the target security level in making any recommendation.  As in, it might be that a 256-bit hash suffices in some cases, but not others.  I think that is secondary to what I think we need to be saying here, but it's probably worth including something.

> The "standard" algorithms listed in this document are suitable for many purposes, including adversarial situations where hash functions might need to provide resistance to collision, first-preimage and second-preimage attacks.  Algorithms listed as "insecure" either provide none of these properties, or are known to be weak (see {{NO-MD5}} and {{NO-SHA1}}).
>
> For adversarial situations, which of the "standard" algorithms are acceptable will depend on the level of protection the circumstances demand. As there is no negotiation, endpoints that depend on a (digest|hash) for security will be vulnerable to attacks on the weakest algorithm they are willing to accept.

**Comment by ioggstream** [comment]
Thanks @martinthomson! @jricher can you PTAL?

**Review by LPardue — APPROVED** [review]
LGTM, thank you all!

File: draft-ietf-httpbis-digest-headers.md [hunk 1]
```diff
@@ -627,8 +627,17 @@ To help endpoints distinguish weaker algorithms from stronger ones,
 this document adds to the IANA Digest Algorithm Values registry
 a new "Status" field containing the most recent appraisal of the digest-algorithm.
 
-An endpoint might have a preference for algorithms,
-such as preferring "standard" algorithms over "insecure" ones.
+The "standard" algorithms listed in this document are suitable for many purposes,
+including adversarial situations where hash functions might need
+to provide resistance to collision, first-preimage and second-preimage attacks.
+Algorithms listed as "insecure" either provide none of these properties,
+or are known to be weak (see {{NO-MD5}} and {{NO-SHA1}}).
```
Comment (LPardue) [inline draft-ietf-httpbis-digest-headers.md:634]:
```suggestion
or are known to be weak (see {{NO-MD5}} and {{NO-SHA}}).
```

ref typo causing build failure

**Comment by LPardue** [comment]
apologies, the ref is `NO-SHA` in main. I changed that to avoid table width alignment problems. if you rebase, it should work. If you need to rename `NO-SHA` to `NO-SHA1` that is fine too.

**Pull request merged**